File: agorachat/src/main/java/io/agora/chatdemo/general/db/AppDatabase.java
package io.agora.chatdemo.general.db;

import androidx.room.Database;
import androidx.room.RoomDatabase;
import androidx.room.TypeConverters;

import io.agora.chatdemo.general.db.converter.DateConverter;
import io.agora.chatdemo.general.db.dao.EmUserDao;
import io.agora.chatdemo.general.db.dao.InviteMessageDao;
import io.agora.chatdemo.general.db.dao.MsgTypeManageDao;
import io.agora.chatdemo.general.db.entity.EmUserEntity;
import io.agora.chatdemo.general.db.entity.InviteMessage;
import io.agora.chatdemo.general.db.entity.MsgTypeManageEntity;


@Database(entities = {EmUserEntity.class,
        InviteMessage.class,
        MsgTypeManageEntity.class,
        },
        version = 1)
@TypeConverters(DateConverter.class)
public abstract class AppDatabase extends RoomDatabase {

    public abstract EmUserDao userDao();

    public abstract InviteMessageDao inviteMessageDao();

    public abstract MsgTypeManageDao msgTypeManageDao();
}

File: app-kotlin/src/main/kotlin/io/agora/chatdemo/common/room/AppDatabase.kt
package io.agora.chatdemo.common.room

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import io.agora.chatdemo.BuildConfig
import io.agora.chatdemo.common.extensions.MD5
import io.agora.chatdemo.common.room.dao.DemoUserDao
import io.agora.chatdemo.common.room.entity.DemoUser

@Database(entities = [DemoUser::class], version = 3)
abstract class AppDatabase: RoomDatabase() {

    /**
     * Get the user data access object.
     */
    abstract fun userDao(): DemoUserDao

    companion object {
        @Volatile
        private var INSTANCE: AppDatabase? = null

        // 以下数据库升级设置，为升级数据库将清掉之前的数据，如果要保留数据，慎重采用此种方式
        // 可以采用addMigrations()的方式，进行数据库的升级
        fun getDatabase(context: Context, userId: String): AppDatabase {
            return INSTANCE ?: synchronized(this) {
                val dbName = (BuildConfig.AGORA_CHAT_APPKEY + userId).MD5()
                val instance = Room.databaseBuilder(
                        context.applicationContext,
                        AppDatabase::class.java,
                        dbName
                    )
                    .allowMainThreadQueries()
                    .fallbackToDestructiveMigration()
                    .build()
                INSTANCE = instance
                instance
            }
        }
    }
}